File: impl/kotlin/src/test/kotlin/HazardKatTest.kt
// Copyright (c) 2026 BubbleFish Technologies, Inc. Apache-2.0.
package sh.bubblefish.naalp

import java.io.File

/**
 * Manufacturing Add-ons Component F (naalp-hazard) known-answer test for the Kotlin SDK, graded
 * against the independent, non-circular oracle `vectors/hazard/cases.json`
 * (tools/hazard_oracle.py) -- mirroring impl/rust/naalp-hazard/src/lib.rs's test module and
 * impl/csharp/test/HazardKatTest.cs, i.e. Kotlin == Rust == C# == oracle.
 *
 * Properties covered (mutation anchors noted):
 *  1. FROM-CODE FAIL-CLOSED -- every from_code row matches the oracle; absent/unknown/out-of-range
 *     input normalizes to MOTION_IN_SHARED_SPACE (F2); the four in-range codes decode to
 *     themselves.
 *  2. BYTE-LEVEL -- claim/authorization encoding and content id match the oracle's body/content-id
 *     hex for every `bodies` row.
 *  3. ROUND-TRIP -- fromValue(toValue(x)) reproduces x's class and bytes for every `bodies` row.
 *  4. COVERAGE MATRIX [MUTATION ANCHOR] -- hazardAuthorized reproduces the oracle's allow/deny
 *     verdict for every `coverage` row; a constant "always authorized" fails every deny row, a
 *     constant "always denied" fails every allow row.
 *  5. ABSENT CLAIM -- hazardAuthorizedOptional(null, grant) denies with the distinct HazardUnknown
 *     error, never HazardNotCovered; a present, well-covered claim still authorizes through the
 *     same entry point.
 *  6. MALFORMED BODIES REJECTED -- fail-closed on empty axes, min > max, non-NFC frame, wrong
 *     shape, a partial body, and an explicit out-of-range class ON THE WIRE (distinct from the
 *     absent/unknown fail-closed normalization path).
 *  7. CONTAINMENT TRUTH TABLES -- spatialContained / envelopeContained direct assertions
 *     independent of the oracle file.
 *
 * KAT convention: a standalone main() exiting non-zero on any failure (mirrors ChannelsKatTest.kt
 * / ApprovalKatTest.kt / RotationKatTest.kt); the filename carries the "KatTest" token the
 * ten-language-parity gate indexes.
 *
 * Run (from impl/kotlin/):
 *   kotlinc src/main/kotlin src/test/kotlin/HazardKatTest.kt -include-runtime -d out/hazard-kat.jar
 *   java -cp out/hazard-kat.jar sh.bubblefish.naalp.HazardKatTestKt
 */

// ---- balanced-brace / regex JSON access (string-aware; mirrors ChannelsKatTest.kt) ----------------

private class HzFailure(msg: String) : RuntimeException(msg)

private var hzFails = 0

private fun hzOk(name: String) = println("  ok   $name")

private fun hzFail(name: String, detail: String) {
    hzFails++
    println("  FAIL $name\n       $detail")
}

private fun hzCheck(name: String, got: String, want: String) {
    if (got == want) hzOk(name) else hzFail(name, "got  $got\n       want $want")
}

private fun hzCheckBool(name: String, got: Boolean, want: Boolean) {
    if (got == want) hzOk(name) else hzFail(name, "got $got want $want")
}

/** Run [block]; require it throws a [NaalpException] whose kind == [wantKind]. */
private fun hzExpectThrows(name: String, wantKind: String, block: () -> Unit) {
    try {
        block()
        hzFail(name, "expected $wantKind, no exception thrown")
    } catch (e: NaalpException) {
        hzCheck(name, e.kind, wantKind)
    }
}

/** Run [block]; require it does NOT throw. */
private fun hzExpectOk(name: String, block: () -> Unit) {
    try {
        block()
        hzOk(name)
    } catch (e: NaalpException) {
        hzFail(name, "expected no error, got ${e.kind}: ${e.message}")
    }
}

private fun bracketExtent(s: String, openCh: Char, closeCh: Char, start: Int): Int {
    var depth = 0
    var inString = false
    var escape = false
    for (i in start until s.length) {
        val c = s[i]
        if (inString) {
            when {
                escape -> escape = false
                c == '\\' -> escape = true
                c == '"' -> inString = false
            }
            continue
        }
        when {
            c == '"' -> inString = true
            c == openCh -> depth++
            c == closeCh -> {
                depth--
                if (depth == 0) return i
            }
        }
    }
    return -1
}

/** The raw substring (including delimiters) of the value following "key": in [s], for an object
 *  ("{...}") or array ("[...]") value. */
private fun rawValue(s: String, key: String): String? {
    val idx = s.indexOf("\"$key\"")
    if (idx < 0) return null
    val after = s.substring(idx)
    val colon = after.indexOf(':')
    if (colon < 0) return null
    val rest = after.substring(colon + 1).trimStart()
    if (rest.isEmpty()) return null
    val opener = rest[0]
    val closer = when (opener) {
        '{' -> '}'
        '[' -> ']'
        else -> return null
    }
    val end = bracketExtent(rest, opener, closer, 0)
    if (end < 0) return null
    return rest.substring(0, end + 1)
}

private fun splitJsonObjects(body: String): List<String> {
    val out = ArrayList<String>()
    var depth = 0
    var start = -1
    var inString = false
    var escape = false
    for (i in body.indices) {
        val c = body[i]
        if (inString) {
            when {
                escape -> escape = false
                c == '\\' -> escape = true
                c == '"' -> inString = false
            }
            continue
        }
        when (c) {
            '"' -> inString = true
            '{' -> {
                if (depth == 0) start = i
                depth++
            }
            '}' -> {
                depth--
                if (depth == 0 && start >= 0) {
                    out.add(body.substring(start, i + 1))
                    start = -1
                }
            }
        }
    }
    return out
}

/** The list of raw JSON object strings inside "key": [ {...}, {...}, ... ]. */
private fun objectArrayField(s: String, key: String): List<String> {
    val raw = rawValue(s, key) ?: return emptyList()
    if (!raw.startsWith("[")) return emptyList()
    return splitJsonObjects(raw.substring(1, raw.length - 1))
}

private fun nullableField(s: String, key: String): String? {
    val m = Regex("\"" + Regex.escape(key) + "\"\\s*:\\s*(null|\"(?:[^\"\\\\]|\\\\.)*\"|-?[0-9]+|true|false)").find(s) ?: return null
    val raw = m.groupValues[1]
    if (raw == "null") return null
    return if (raw.startsWith("\"")) raw.substring(1, raw.length - 1) else raw
}

/** Decode JSON string escapes (`\"`, `\\`, `\/`, `\b`, `\f`, `\n`, `\r`, `\t`, `\uXXXX`) -- the
 *  oracle's `unicode_frame` case carries a `é` escape, which a naive quote-strip must not
 *  pass through literally as the six-character text `é`. */
private fun jsonUnescape(s: String): String {
    val sb = StringBuilder(s.length)
    var i = 0
    while (i < s.length) {
        val c = s[i]
        if (c == '\\' && i + 1 < s.length) {
            when (val e = s[i + 1]) {
                '"' -> { sb.append('"'); i += 2 }
                '\\' -> { sb.append('\\'); i += 2 }
                '/' -> { sb.append('/'); i += 2 }
                'b' -> { sb.append('\b'); i += 2 }
                'f' -> { sb.append('\u000C'); i += 2 }
                'n' -> { sb.append('\n'); i += 2 }
                'r' -> { sb.append('\r'); i += 2 }
                't' -> { sb.append('\t'); i += 2 }
                'u' -> {
                    if (i + 6 <= s.length) {
                        sb.append(s.substring(i + 2, i + 6).toInt(16).toChar())
                        i += 6
                    } else {
                        sb.append(e); i += 2
                    }
                }
                else -> { sb.append(e); i += 2 }
            }
        } else {
            sb.append(c)
            i += 1
        }
    }
    return sb.toString()
}

/** A quoted-string field, JSON-unescaped (unlike [nullableField], which leaves escapes raw for
 *  the numeric/bool/null callers that never need them). */
private fun strField(s: String, key: String): String {
    val m = Regex("\"" + Regex.escape(key) + "\"\\s*:\\s*\"((?:[^\"\\\\]|\\\\.)*)\"").find(s)
        ?: throw HzFailure("string key not found: $key")
    return jsonUnescape(m.groupValues[1])
}

/** An unsigned 64-bit-safe numeric field: parses via [java.lang.Long.parseUnsignedLong] so a value
 *  up to 2^64-1 -- carried as either a JSON number or (past 2^63-1) a decimal string in the oracle
 *  -- decodes to the exact bit pattern [Cbor.U] carries, never overflowing/throwing the way a plain
 *  signed `.toLong()` would (mirrors RecheckSignerCounterKatTest.kt's uintField). */
private fun uField(s: String, key: String): Long =
    nullableField(s, key)?.let { java.lang.Long.parseUnsignedLong(it) } ?: throw HzFailure("uint key not found: $key")

private fun nullableUField(s: String, key: String): Long? =
    nullableField(s, key)?.let { java.lang.Long.parseUnsignedLong(it) }

/** The list of `[min, max]` signed-integer pairs inside "axes": [ [a,b], [c,d], ... ] -- a flat
 *  regex scan suffices because axes never nest a string or another array inside a pair. */
private fun axesFrom(s: String): List<Pair<Long, Long>> {
    val raw = rawValue(s, "axes") ?: throw HzFailure("missing axes")
    return Regex("\\[\\s*(-?\\d+)\\s*,\\s*(-?\\d+)\\s*\\]").findAll(raw)
        .map { m -> m.groupValues[1].toLong() to m.groupValues[2].toLong() }
        .toList()
}

private fun envFrom(o: String): Hazard.HazardEnvelope {
    val frame = strField(o, "frame")
    val axes = axesFrom(o)
    val speed = uField(o, "speed_bound_mm_s")
    val notBefore = uField(o, "not_before")
    val notAfter = uField(o, "not_after")
    return Hazard.HazardEnvelope(Hazard.SpatialBounds(frame, axes), speed, Hazard.HazardWindow(notBefore, notAfter))
}

private fun findVector(rel: String): File {
    var d: File? = File(".").absoluteFile
    repeat(8) {
        val cur = d ?: throw HzFailure("$rel not found")
        val p = File(cur, rel)
        if (p.isFile) return p
        d = cur.parentFile
    }
    throw HzFailure("$rel not found from ${File(".").absolutePath}")
}

private fun vectorJson(): String = findVector("vectors/hazard/cases.json").readText(Charsets.UTF_8)

// ==== 1. FROM-CODE FAIL-CLOSED (F2) matches the oracle ==============================================

private fun fromCodeFailClosedRun() {
    val c = vectorJson()
    for (row in objectArrayField(c, "from_code")) {
        val input = nullableUField(row, "code")
        val want = uField(row, "class")
        hzCheck("from_code($input)", Hazard.HazardClass.fromCode(input).code.toString(), want.toString())
    }

    // Explicit oracle-independent assertions of the two named fail-closed cases (F2).
    hzCheckBool("from_code(null) == MOTION_IN_SHARED_SPACE",
        Hazard.HazardClass.fromCode(null) == Hazard.HazardClass.MOTION_IN_SHARED_SPACE, true)
    hzCheckBool("from_code(9) == MOTION_IN_SHARED_SPACE",
        Hazard.HazardClass.fromCode(9L) == Hazard.HazardClass.MOTION_IN_SHARED_SPACE, true)
    // u64::MAX (0xFFFFFFFFFFFFFFFF) as its Long bit pattern is -1L; an out-of-u8-range code must
    // still normalize, not throw or wrap into an in-range class.
    hzCheckBool("from_code(u64::MAX) == MOTION_IN_SHARED_SPACE",
        Hazard.HazardClass.fromCode(-1L) == Hazard.HazardClass.MOTION_IN_SHARED_SPACE, true)

    // The five in-range codes decode to themselves, never collapsing to the default.
    for (code in 0L..4L) {
        hzCheck("from_code($code) round-trips its own code", Hazard.HazardClass.fromCode(code).code.toString(), code.toString())
    }
}

// ==== 2/3. BYTE-LEVEL + ROUND-TRIP: encode matches the independent oracle ============================

private fun bytesAndRoundTripRun() {
    val c = vectorJson()
    for (row in objectArrayField(c, "bodies")) {
        val name = strField(row, "name")
        val cls = Hazard.HazardClass.fromCode(uField(row, "class"))
        val e = envFrom(row)
        val claim = Hazard.HazardClaim(cls, e)
        val auth = Hazard.HazardAuthorization(cls, e)
        val wantBody = strField(row, "body_hex")
        hzCheck("$name claim bytes == oracle body_hex", Hex.encode(claim.bytes()), wantBody)
        hzCheck("$name authorization bytes == oracle body_hex (same shape as claim)", Hex.encode(auth.bytes()), wantBody)
        hzCheck("$name claim content-id == oracle", Hex.encode(claim.contentId()), strField(row, "content_id_hex"))

        val got = Hazard.HazardClaim.fromValue(claim.toValue())
        hzCheckBool("$name round-trip class", got.hazardClass == claim.hazardClass, true)
        hzCheck("$name round-trip bytes", Hex.encode(got.bytes()), Hex.encode(claim.bytes()))
    }
}

// ==== 4. COVERAGE MATRIX [MUTATION ANCHOR] ============================================================
//
// A constant "always authorized" hazardAuthorized fails every deny row below; a constant "always
// denied" fails every allow row. The matrix needs both to be a real mutation anchor.

private fun coverageMatchesOracleRun() {
    val c = vectorJson()
    val rows = objectArrayField(c, "coverage")
    hzCheckBool("coverage matrix non-empty", rows.isNotEmpty(), true)
    var allows = 0
    var denies = 0
    for (row in rows) {
        val name = strField(row, "name")
        val claim = Hazard.HazardClaim(
            Hazard.HazardClass.fromCode(nullableUField(row, "claim_class_code")),
            envFrom(rawValue(row, "claim_envelope")!!),
        )
        val grant = Hazard.HazardAuthorization(
            Hazard.HazardClass.fromCode(nullableUField(row, "grant_class_code")),
            envFrom(rawValue(row, "grant_envelope")!!),
        )
        val wantOk = nullableField(row, "authorized") == "true"
        if (wantOk) {
            allows++
            hzExpectOk("$name authorized") { Hazard.hazardAuthorized(claim, grant) }
        } else {
            denies++
            hzExpectThrows("$name denied", "HazardNotCovered") { Hazard.hazardAuthorized(claim, grant) }
        }
    }
    hzCheckBool("matrix needs both allows and denies", allows > 0 && denies > 0, true)
}

// ==== 5. ABSENT CLAIM: distinct HazardUnknown, never HazardNotCovered ================================
//
// F2/F4's "absent hazard" behavioural vector, at the OBJECT level (no claim at all): distinct from
// an in-range-but-mismatched class, and distinct from an unrecognized class byte inside a present
// claim (covered by coverageMatchesOracleRun's normalized rows).

private fun absentClaimDeniesWithDistinctErrorRun() {
    val grant = Hazard.HazardAuthorization(
        Hazard.HazardClass.TOOL_ACTUATION,
        Hazard.HazardEnvelope(
            Hazard.SpatialBounds("cell-7/world", listOf(0L to 1000L, 0L to 1000L, 0L to 500L)),
            500L, Hazard.HazardWindow(0L, 1000L),
        ),
    )
    hzExpectThrows("absent claim denies with HazardUnknown", "HazardUnknown") {
        Hazard.hazardAuthorizedOptional(null, grant)
    }
    // A present, well-covered claim still authorizes through the same entry point.
    val claim = Hazard.HazardClaim(
        Hazard.HazardClass.TOOL_ACTUATION,
        Hazard.HazardEnvelope(
            Hazard.SpatialBounds("cell-7/world", listOf(100L to 200L, 100L to 200L, 0L to 100L)),
            100L, Hazard.HazardWindow(10L, 900L),
        ),
    )
    hzExpectOk("present well-covered claim authorizes via the optional entry point") {
        Hazard.hazardAuthorizedOptional(claim, grant)
    }
}

// ==== 6. MALFORMED BODIES REJECTED (fail-closed, never partially valid) ==============================

private fun malformedBodiesRejectedRun() {
    // empty axes
    val bad = Hazard.SpatialBounds("f", emptyList())
    hzCheckBool("empty-axes SpatialBounds is not well-formed", bad.isWellFormed(), false)
    hzExpectThrows("empty-axes SpatialBounds.fromValue rejected", "HazardMalformed") {
        Hazard.SpatialBounds.fromValue(bad.toValue())
    }

    // min > max
    val bad2 = Hazard.SpatialBounds("f", listOf(10L to -10L))
    hzCheckBool("min>max SpatialBounds is not well-formed", bad2.isWellFormed(), false)

    // non-NFC frame ('e' + combining acute, NFD not NFC)
    val bad3 = Hazard.SpatialBounds("é", listOf(0L to 1L))
    hzCheckBool("non-NFC frame SpatialBounds is not well-formed", bad3.isWellFormed(), false)

    // wrong shape entirely (not a map)
    hzExpectThrows("non-map HazardClaim.fromValue rejected", "HazardMalformed") {
        Hazard.HazardClaim.fromValue(Cbor.U(0))
    }

    // class present, envelope missing
    val partial = Cbor.M(listOf(Cbor.Pair(Cbor.U(1), Cbor.U(1))))
    hzExpectThrows("class-only body rejected", "HazardMalformed") {
        Hazard.HazardClaim.fromValue(partial)
    }

    // an explicit out-of-range class ON THE WIRE (not "absent") is malformed, not silently
    // normalized -- see Hazard.hazardBodyFromValue's doc comment.
    val goodEnv = Hazard.HazardEnvelope(Hazard.SpatialBounds("f", listOf(0L to 1L)), 1L, Hazard.HazardWindow(0L, 1L))
    val outOfRange = Cbor.M(listOf(
        Cbor.Pair(Cbor.U(1), Cbor.U(99)),
        Cbor.Pair(Cbor.U(2), goodEnv.toValue()),
    ))
    hzExpectThrows("out-of-range class on the wire rejected (not normalized)", "HazardMalformed") {
        Hazard.HazardClaim.fromValue(outOfRange)
    }
}

// ==== 7. CONTAINMENT TRUTH TABLES (independent of the oracle file, direct assertions) ================

private fun spatialContainedTruthTableRun() {
    val grant = Hazard.SpatialBounds("f", listOf(0L to 100L, 0L to 100L))
    // fully inside -> contained
    val inside = Hazard.SpatialBounds("f", listOf(10L to 90L, 10L to 90L))
    hzCheckBool("fully-inside contained", Hazard.spatialContained(inside, grant), true)
    // equal bounds -> contained (closed interval)
    val equal = Hazard.SpatialBounds("f", listOf(0L to 100L, 0L to 100L))
    hzCheckBool("equal-bounds contained (closed interval)", Hazard.spatialContained(equal, grant), true)
    // one axis pokes outside -> not contained
    val outside = Hazard.SpatialBounds("f", listOf(10L to 90L, 10L to 101L))
    hzCheckBool("one-axis-outside not contained", Hazard.spatialContained(outside, grant), false)
    // different frame -> never contained regardless of numeric bounds
    val wrongFrame = Hazard.SpatialBounds("g", listOf(10L to 90L, 10L to 90L))
    hzCheckBool("different-frame never contained", Hazard.spatialContained(wrongFrame, grant), false)
    // fewer axes -> never contained
    val fewer = Hazard.SpatialBounds("f", listOf(10L to 90L))
    hzCheckBool("fewer-axes never contained", Hazard.spatialContained(fewer, grant), false)
}

private fun envelopeContainedWindowAndSpeedRun() {
    fun env(lo: Long, hi: Long, speed: Long, nb: Long, na: Long) =
        Hazard.HazardEnvelope(Hazard.SpatialBounds("f", listOf(lo to hi)), speed, Hazard.HazardWindow(nb, na))

    val grant = env(0, 100, 500, 100, 900)
    val ok = env(0, 100, 500, 100, 900) // exact edges, closed interval
    hzCheckBool("exact-edges envelope contained", Hazard.envelopeContained(ok, grant), true)
    val speedOver = env(0, 100, 501, 100, 900)
    hzCheckBool("speed-over-bound not contained", Hazard.envelopeContained(speedOver, grant), false)
    val startsEarly = env(0, 100, 500, 99, 900)
    hzCheckBool("window-starts-early not contained", Hazard.envelopeContained(startsEarly, grant), false)
    val endsLate = env(0, 100, 500, 100, 901)
    hzCheckBool("window-ends-late not contained", Hazard.envelopeContained(endsLate, grant), false)
}

fun main() {
    println("HazardKatTest (Kotlin) -- Manufacturing Add-ons Component F (naalp-hazard)")
    fromCodeFailClosedRun()
    bytesAndRoundTripRun()
    coverageMatchesOracleRun()
    absentClaimDeniesWithDistinctErrorRun()
    malformedBodiesRejectedRun()
    spatialContainedTruthTableRun()
    envelopeContainedWindowAndSpeedRun()
    println(if (hzFails == 0) "HazardKatTest: PASS" else "HazardKatTest: FAIL ($hzFails)")
    if (hzFails != 0) kotlin.system.exitProcess(1)
}

File: impl/php/test/hazard_kat_test.php
<?php
// Copyright (c) 2026 BubbleFish Technologies, Inc. Apache-2.0.
//
// Manufacturing Add-ons Component F (naalp-hazard) known-answer test for the PHP SDK, graded
// against the independent, non-circular oracle vectors/hazard/cases.json (tools/hazard_oracle.py)
// -- mirroring impl/rust/naalp-hazard/src/lib.rs's test module and impl/csharp/test/HazardKatTest.cs
// / impl/java/.../HazardKatTest.java, i.e. PHP == Rust == oracle.
//
// Run: php -d extension=sodium -d extension=intl -d ffi.enable=1 -d memory_limit=-1
//          test/hazard_kat_test.php   (from impl/php/). Exit code 0 = all checks passed.

declare(strict_types=1);

require __DIR__ . '/../src/bootstrap.php';

use Naalp\Hazard;
use Naalp\HazardAuthorization;
use Naalp\HazardClaim;
use Naalp\HazardEnvelope;
use Naalp\HazardWindow;
use Naalp\SpatialBounds;
use Naalp\Cbor;
use Naalp\U;

$fails = 0;
function check(string $name, bool $ok, string $detail = ""): void
{
    global $fails;
    if ($ok) {
        echo "  ok   $name\n";
    } else {
        $fails++;
        echo "  FAIL $name" . ($detail !== "" ? "  $detail" : "") . "\n";
    }
}

/** Run $fn and return the caught error's ->kind (or "no-error" / the class name). */
function err_kind(callable $fn): string
{
    try {
        $fn();
        return "no-error";
    } catch (\Throwable $e) {
        return \property_exists($e, 'kind') ? $e->kind : \get_class($e);
    }
}

/** Walk up from this dir to the repository's shared corpus at $rel (the independent oracle). */
function shared_vectors(string $rel): array
{
    $d = __DIR__;
    for ($i = 0; $i < 6; $i++) {
        $p = $d . '/vectors/' . $rel;
        if (\is_file($p)) {
            return \json_decode(\file_get_contents($p), true, 512, JSON_THROW_ON_ERROR);
        }
        $d = \dirname($d);
    }
    throw new \RuntimeException("vectors/$rel not found");
}

/**
 * A `from_code` row's "code" cell is null, a JSON number, or (for the > 2^63-1 case) a decimal
 * string -- never a float64 decoder anywhere, so the exact wire value survives. Convert the
 * decimal string to the SAME two's-complement uint64 bit-pattern PHP int that Naalp\U::$v uses
 * (Cbor.php's own convention: a value >= 2^63 is a negative PHP int), via bcmath so no precision
 * is lost between "2^63" and "PHP_INT_MAX".
 */
function code_from_json(mixed $v): ?int
{
    if ($v === null) {
        return null;
    }
    if (\is_int($v)) {
        return $v;
    }
    // decimal string, in [0, 2^64-1]
    if (\bccomp($v, '9223372036854775807') <= 0) {
        return (int) $v;
    }
    return (int) \bcsub($v, '18446744073709551616'); // wrap into the signed 64-bit bit pattern
}

function axes_from(array $axesEl): array
{
    $out = [];
    foreach ($axesEl as $pair) {
        $out[] = [(int) $pair[0], (int) $pair[1]];
    }
    return $out;
}

function env_from(array $o): HazardEnvelope
{
    return new HazardEnvelope(
        new SpatialBounds((string) $o['frame'], axes_from($o['axes'])),
        (int) $o['speed_bound_mm_s'],
        new HazardWindow((int) $o['not_before'], (int) $o['not_after']),
    );
}

echo "naalp-hazard KAT (PHP, Manufacturing Add-ons Component F)\n";

$c = shared_vectors('hazard/cases.json');

// ---------------------------------------------------------------------------------------------
// F2: fail-closed class decode (mutation anchor: a constant CLASS_NONE return would pass none of
// the non-zero cases; a constant CLASS_MOTION_IN_SHARED_SPACE would fail the exact 0..3 cases).
// ---------------------------------------------------------------------------------------------
echo "\n1. from_code fail-closed decode\n";

foreach ($c['from_code'] as $row) {
    $input = code_from_json($row['code']);
    $want = (int) $row['class'];
    $got = Hazard::classFromCode($input);
    check("from_code(" . ($input === null ? "null" : $input) . ") == $want ({$row['class_name']})",
        $got === $want, "got $got");
}

// Explicit oracle-independent assertions of the two named fail-closed cases (F2).
check("absent must normalize to the highest class",
    Hazard::classFromCode(null) === Hazard::CLASS_MOTION_IN_SHARED_SPACE);
check("unknown code must normalize to the highest class",
    Hazard::classFromCode(9) === Hazard::CLASS_MOTION_IN_SHARED_SPACE);
check("an out-of-range code (PHP_INT_MAX bit pattern) must still normalize, not throw",
    Hazard::classFromCode(\PHP_INT_MAX) === Hazard::CLASS_MOTION_IN_SHARED_SPACE);

// The five in-range codes decode to themselves, never collapsing to the default.
for ($code = 0; $code <= 4; $code++) {
    check("from_code($code) round-trips to itself", Hazard::classFromCode($code) === $code);
}

// ---------------------------------------------------------------------------------------------
// byte-level: encode matches the independent oracle (=> PHP == Rust == C# == Java once graded).
// ---------------------------------------------------------------------------------------------
echo "\n2. claim/authorization bytes match oracle\n";

foreach ($c['bodies'] as $row) {
    $class = Hazard::classFromCode((int) $row['class']);
    $e = env_from($row);
    $claim = new HazardClaim($class, $e);
    $auth = new HazardAuthorization($class, $e);
    $name = $row['name'];
    $want = $row['body_hex'];
    check("claim $name body_hex", \bin2hex($claim->bytes()) === $want,
        "got " . \bin2hex($claim->bytes()));
    check("authorization $name body_hex (same shape as claim)", \bin2hex($auth->bytes()) === $want,
        "got " . \bin2hex($auth->bytes()));
    check("claim $name content_id_hex", \bin2hex($claim->contentId()) === $row['content_id_hex'],
        "got " . \bin2hex($claim->contentId()));
}

// ---------------------------------------------------------------------------------------------
// Round-trip: fromValue(toValue(x)) == x for every oracle body.
// ---------------------------------------------------------------------------------------------
echo "\n3. round-trip matches oracle\n";

foreach ($c['bodies'] as $row) {
    $class = Hazard::classFromCode((int) $row['class']);
    $e = env_from($row);
    $claim = new HazardClaim($class, $e);
    $got = HazardClaim::fromValue($claim->toValue());
    check("round-trip {$row['name']} class", $got->class === $claim->class);
    check("round-trip {$row['name']} bytes", \bin2hex($got->bytes()) === \bin2hex($claim->bytes()));
}

// ---------------------------------------------------------------------------------------------
// F3: coverage matrix (mutation anchor: a constant "always authorized" fails the deny rows; a
// constant "always denied" fails the allow rows).
// ---------------------------------------------------------------------------------------------
echo "\n4. coverage matches oracle\n";

$rows = $c['coverage'];
check("coverage matrix is non-empty", \count($rows) > 0);
$allows = 0;
$denies = 0;
foreach ($rows as $row) {
    $claim = new HazardClaim(
        Hazard::classFromCode(code_from_json($row['claim_class_code'])),
        env_from($row['claim_envelope']),
    );
    $grant = new HazardAuthorization(
        Hazard::classFromCode(code_from_json($row['grant_class_code'])),
        env_from($row['grant_envelope']),
    );
    $wantOk = (bool) $row['authorized'];
    $name = $row['name'];
    if ($wantOk) {
        $allows++;
        check("coverage $name -> authorized", err_kind(fn() => Hazard::hazardAuthorized($claim, $grant)) === "no-error");
    } else {
        $denies++;
        check("coverage $name -> HazardNotCovered",
            err_kind(fn() => Hazard::hazardAuthorized($claim, $grant)) === "HazardNotCovered");
    }
}
check("matrix needs both allows and denies", $allows > 0 && $denies > 0, "allows=$allows denies=$denies");

// ---------------------------------------------------------------------------------------------
// F2/F4's "absent hazard" behavioural vector, at the OBJECT level (no claim at all): distinct
// from an in-range-but-mismatched class, and distinct from an unrecognized class byte inside a
// present claim (covered by the coverage matrix's normalized rows).
// ---------------------------------------------------------------------------------------------
echo "\n5. absent claim denies with distinct error\n";

$grant = new HazardAuthorization(
    Hazard::CLASS_TOOL_ACTUATION,
    new HazardEnvelope(
        new SpatialBounds("cell-7/world", [[0, 1000], [0, 1000], [0, 500]]),
        500,
        new HazardWindow(0, 1000),
    ),
);
check("absent claim -> HazardUnknown",
    err_kind(fn() => Hazard::hazardAuthorizedOptional(null, $grant)) === "HazardUnknown");

$claim = new HazardClaim(
    Hazard::CLASS_TOOL_ACTUATION,
    new HazardEnvelope(
        new SpatialBounds("cell-7/world", [[100, 200], [100, 200], [0, 100]]),
        100,
        new HazardWindow(10, 900),
    ),
);
check("present, well-covered claim still authorizes through the same entry point",
    err_kind(fn() => Hazard::hazardAuthorizedOptional($claim, $grant)) === "no-error");

// ---------------------------------------------------------------------------------------------
// structural malformation (fail-closed, never partially valid).
// ---------------------------------------------------------------------------------------------
echo "\n6. malformed bodies rejected\n";

// empty axes
$bad = new SpatialBounds("f", []);
check("empty axes -> not well-formed", !$bad->isWellFormed());
check("empty axes -> HazardMalformed on decode",
    err_kind(fn() => SpatialBounds::fromValue($bad->toValue())) === "HazardMalformed");

// min > max
$bad2 = new SpatialBounds("f", [[10, -10]]);
check("min > max -> not well-formed", !$bad2->isWellFormed());

// non-NFC frame ('e' + combining acute, NFD not NFC)
$bad3 = new SpatialBounds("e\u{0301}", [[0, 1]]);
check("non-NFC frame -> not well-formed", !$bad3->isWellFormed());

// wrong shape entirely (not a map)
check("wrong shape -> HazardMalformed",
    err_kind(fn() => HazardClaim::fromValue(new U(0))) === "HazardMalformed");

// class present, envelope missing
$partial = new Naalp\M([[new U(1), new U(1)]]);
check("class present, envelope missing -> HazardMalformed",
    err_kind(fn() => HazardClaim::fromValue($partial)) === "HazardMalformed");

// an explicit out-of-range class ON THE WIRE (not "absent") is malformed, not silently
// normalized -- see Hazard::bodyFromValue's doc comment.
$goodEnv = (new HazardEnvelope(new SpatialBounds("f", [[0, 1]]), 1, new HazardWindow(0, 1)))->toValue();
$outOfRange = new Naalp\M([[new U(1), new U(99)], [new U(2), $goodEnv]]);
check("out-of-range class on the wire -> HazardMalformed",
    err_kind(fn() => HazardClaim::fromValue($outOfRange)) === "HazardMalformed");

// ---------------------------------------------------------------------------------------------
// containment truth table (independent of the oracle file, direct assertions).
// ---------------------------------------------------------------------------------------------
echo "\n7. spatial_contained truth table\n";

$grantSb = new SpatialBounds("f", [[0, 100], [0, 100]]);
$inside = new SpatialBounds("f", [[10, 90], [10, 90]]);
check("fully inside -> contained", Hazard::spatialContained($inside, $grantSb));
$equal = new SpatialBounds("f", [[0, 100], [0, 100]]);
check("equal bounds -> contained (closed interval)", Hazard::spatialContained($equal, $grantSb));
$outside = new SpatialBounds("f", [[10, 90], [10, 101]]);
check("one axis pokes outside -> not contained", !Hazard::spatialContained($outside, $grantSb));
$wrongFrame = new SpatialBounds("g", [[10, 90], [10, 90]]);
check("different frame -> never contained regardless of numeric bounds", !Hazard::spatialContained($wrongFrame, $grantSb));
$fewer = new SpatialBounds("f", [[10, 90]]);
check("fewer axes -> never contained", !Hazard::spatialContained($fewer, $grantSb));

echo "\n8. envelope_contained window and speed\n";

$mkEnv = fn(int $lo, int $hi, int $speed, int $nb, int $na) =>
    new HazardEnvelope(new SpatialBounds("f", [[$lo, $hi]]), $speed, new HazardWindow($nb, $na));

$envGrant = $mkEnv(0, 100, 500, 100, 900);
$ok = $mkEnv(0, 100, 500, 100, 900); // exact edges, closed interval
check("exact edges -> contained", Hazard::envelopeContained($ok, $envGrant));
$speedOver = $mkEnv(0, 100, 501, 100, 900);
check("speed over bound -> not contained", !Hazard::envelopeContained($speedOver, $envGrant));
$startsEarly = $mkEnv(0, 100, 500, 99, 900);
check("window starts before grant -> not contained", !Hazard::envelopeContained($startsEarly, $envGrant));
$endsLate = $mkEnv(0, 100, 500, 100, 901);
check("window ends after grant -> not contained", !Hazard::envelopeContained($endsLate, $envGrant));

echo "\n" . ($fails === 0 ? "PASS" : "FAIL ($fails)") . "\n";
exit($fails === 0 ? 0 : 1);
